Hierachical projpred

AlejandroCatalina opened the pull request:
Here we are! This is a big PR so it will take some time to go through it but we are finally bringing projpred support to multilevel models. 

The good thing is that the interface for simpler models shouldn't change with respect to the current one so most use cases will remain unchanged.

paul-buerkner commented:
Thank you, @AlejandroCatalina, for refactoring and extending projpred!

jpiironen commented:
Thanks @AlejandroCatalina ! I will go through some of the stuff to see if I have some comments, but it will take a while as this is a huge PR and I haven't been working the codebase for a while.

AlejandroCatalina commented:
Yes for sure.

The general structure is the same, we just simplified most of the stuff by outsourcing the key functionality for the projection. Beware that for multilevel models we can only do forward search and therefore that method is a bit heavier now to handle the search. But of course, I will gladly address any issues/questions/comments you have!

Best,
Alejandro
On 18 Feb 2020, 18:51 +0200, Juho Piironen <notifications@github.com>, wrote:
> Thanks @AlejandroCatalina ! I will go through some of the stuff to see if I have some comments, but it will take a while as this is a huge PR and I haven't been working the codebase for a while.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jpiironen commented on R/search_formula.R:
```diff
@@ -0,0 +1,84 @@
+  spath <- search_L1(p_ref, list(refmodel, x=x), family, intercept, nv_max, penalty, opt)
+  ## TODO: check this? can we reduce the above line and this one to a single thing?
+  ## extract the path from glmnet
+  sub_fit <- glmnet::glmnet(x, p_ref$mu, family, intercept = intercept, dfmax = nv_max, lambda = penalty)
```
Is this really necessary? I believe you should be able to get the same information from search_L1 because it computes the whole L1-regularization path. If `search_L1` does not return exactly what you need, I guess it could be made to do so with small changes. 

I think the current version is dangerous, because I'm not convinced at all that the lambda-values with our L1-computation and that of glmnet match each other exactly, and hence there is a lot of potential for error here. Another reason is that I would like to avoid dependency to glmnet if this is only place where it's used, because then we will be restricted to those families only that can be found in glmnet.

AlejandroCatalina replied:
I agree in that regard, I set the TODO to work this out. The reason why I return the sub_fit is because it resembles the same structure for multilevel models, where each of the sub_fits can be called to make predictions. I think a nicer workaround for this is to come up with a small minimal structure that mimics a model and contains alpha and beta basically to make predictions. Would that be okay from your perspective?

AlejandroCatalina commented:
I should say as well, @jpiironen, that @paul-buerkner has already agreed to do a review, in case you are too busy to take care of it. But of course the more reviews the better as this is implies a big change over the current codebase.

jpiironen commented:
> I should say as well, @jpiironen, that @paul-buerkner has already agreed to do a review, in case you are too busy to take care of it. But of course the more reviews the better as this is implies a big change over the current codebase.

Sure, I did not even intend to do an exhaustive review, I simply planned to glance through the critical parts regarding the methodology code to see if I have something useful to say about those. I trust that @paul-buerkner can review the code quality and design more generally. :)

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+cv_varsel_poc <- function(fit,  method = NULL, cv_method = NULL,
```
copied documentation over from the current version

AlejandroCatalina commented on DESCRIPTION:
```diff
@@ -1,14 +1,16 @@
 Maintainer:
-    Juho Piironen <juho.t.piironen@gmail.com>
+    Juho Piironen <juho@cai.fi>
+    Alejandro Catalina <alejandro.catalina@aalto.fi>
```
right, I didn't know about this so I just thought that I could help given the many changes introduced, but I will take it back to the current version and let's see

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+                          thresh=1e-6, regul=1e-4, validate_search=T, seed=NULL,
+                          groups=NULL, ...) {
+
+  refmodel <- get_refmodel_poc(fit, ...)
```
As I don't see this on my editor, it's going to be hard to fix. Hopefully it's fixed in the latest commit.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  refmodel <- get_refmodel_poc(fit, ...)
+
+	# resolve the arguments similar to varsel
+	args <- parseargs_varsel_poc(refmodel, method, relax, intercept, nv_max, nc, ns, ncpred, nspred, groups)
```
cool

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  refmodel <- get_refmodel_poc(fit, ...)
+
+	# resolve the arguments similar to varsel
+	args <- parseargs_varsel_poc(refmodel, method, relax, intercept, nv_max, nc, ns, ncpred, nspred, groups)
```
cool

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	args <- parseargs_varsel_poc(refmodel, method, relax, intercept, nv_max, nc, ns, ncpred, nspred, groups)
+	method <- args$method
+	relax <- args$relax
+	intercept <- args$intercept
```
It refers to fitting an intercept in the submodels or not. This is carried over from the current version. I think we have already discussed that we should always include an intercept, so I think we should lean towards removing this?

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	K <- args$K
+
+	# search options
+	opt <- list(lambda_min_ratio=lambda_min_ratio, nlambda=nlambda, thresh=thresh, regul=regul)
```
nice. I'll take a pass on other parts of the package where we use this kind of list structure.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	# search options
+	opt <- list(lambda_min_ratio=lambda_min_ratio, nlambda=nlambda, thresh=thresh, regul=regul)
+
+  if (tolower(cv_method) == 'loo')  {
```
okay

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+
+  if (tolower(cv_method) == 'loo')  {
+    if (!(is.null(K))) warning('K provided, but cv_method is LOO.')
+    sel_cv <- loo_varsel_poc(refmodel, method, nv_max, ns, nc, nspred, ncpred, relax, intercept, penalty,
```
agreed, I took a pass naming the arguments here and in other places.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	ch <- as.matrix(unname(as.data.frame(sel_cv$vind_cv)))
+  w <- sel_cv$summaries$sub[[1]]$w # these weights might be non-constant in case of subsampling LOO
+  selvind <- sel$vind
+  if (is.null(w)) w <- rep(1/ncol(ch), ncol(ch)) # if weights are not set, then all validation folds have equal weight
```
Yes, it depends on the specific CV method used to validate the search. In loo we have psisloo weights, but in k fold might be different (usually null and therefore we use equal weights for all data).

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  vars <- unlist(selvind)
+  pctch <- t(sapply(seq_along(selvind), function(size) {
+    c(size = size, sapply(vars, function(var) {
+      sum(t(ch[1:size, ] == var) * w, na.rm = T)
```
cool

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  vars <- unlist(selvind)
+  pctch <- t(sapply(seq_along(selvind), function(size) {
+    c(size = size, sapply(vars, function(var) {
+      sum(t(ch[1:size, ] == var) * w, na.rm = T)
```
yes

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	vs <- list()
+	vs$refmodel <- refmodel
+	vs$spath <- sel$spath
+	vs <- c(vs, c(sel_cv[c('d_test', 'summaries')],
```
I changed it to list(...) notation.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	              sel[c('family_kl', 'vind', 'kl')],
+	              list(pctch = pctch)))
+  class(vs) <- 'cvsel'
+	vs$nv_max <- nv_max
```
none, carried over from current version, fixed it on the latest commit.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	vs
+}
+
+parseargs_cv_varsel <- function(refmodel, cv_method, K) {
```
added documentation

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+
+parseargs_cv_varsel <- function(refmodel, cv_method, K) {
+
+  ##
```
fixed

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  ## fold (for each data point)
+  ##
+
+  group_features <- !is.null(groups)
```
okay

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  ##
+
+  group_features <- !is.null(groups)
+  fam <- refmodel$fam
```
renamed to family

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  group_features <- !is.null(groups)
+  fam <- refmodel$fam
+  mu <- refmodel$mu
+  dis <- refmodel$dis
```
dis refers to dispersion, whereas dist to distribution, we can make it more verbose and change dis to dispersion

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	cl_pred <- p_pred$cl
+
+	# fetch the log-likelihood for the reference model to obtain the LOO weights
+	if (is.null(refmodel$loglik))
```
added in places where there are comments

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	else
+		# log-likelihood available
+		loglik <- refmodel$loglik
+	psisloo <- loo::psis(-loglik, cores = 1, r_eff = rep(1,ncol(loglik))) # TODO: should take r_eff:s into account
```
I'm not really sure what this todo means either, it's carried over from the current version (at least at the time when I forked the repo)

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	else
+		# log-likelihood available
+		loglik <- refmodel$loglik
+	psisloo <- loo::psis(-loglik, cores = 1, r_eff = rep(1,ncol(loglik))) # TODO: should take r_eff:s into account
```
I don't have a strong opinion. Generally there shouldn't be a need for more than 1 core but if the user wants to use more, so be it.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	lw <- weights(psisloo)
+	pareto_k <- loo::pareto_k_values(psisloo)
+	n <- length(pareto_k)
+	nloo <- ifelse(is.null(nloo), n, nloo) # by default use all observations
```
nice

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	nloo <- min(nloo,n)
+
+	# compute loo summaries for the reference model
+	loo_ref <- apply(loglik+lw, 2, 'log_sum_exp')
```
fixed

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	nloo <- min(nloo,n)
+
+	# compute loo summaries for the reference model
+	loo_ref <- apply(loglik+lw, 2, 'log_sum_exp')
```
fixed

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	if (!validate_search) {
+	  # perform selection only once using all the data (not separately for each fold),
+	  # and perform the projection then for each submodel size
+	  # vind <- select(method, p_sel, refmodel, fam, intercept, nv_max, penalty, verbose=F, opt)
```
removed the commented out code

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+    vind <- spath$vind
+  }
+
+	for (run_index in seq_along(inds)) {
```
this loop runs over the subsampled loo observations to run the search for.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	  # observation index
+	  i <- inds[run_index]
+
+	  # reweight the clusters/samples according to the is-loo weights
```
they are psisloo weights in our implementation always

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	  p_sel <- .get_p_clust(fam, mu, dis, wsample=exp(lw[,i]), cl=cl_sel)
+	  p_pred <- .get_p_clust(fam, mu, dis, wsample=exp(lw[,i]), cl=cl_pred)
+
+		if (validate_search) {
```
Not sure what you mean. We only need to run an extra search if validate_search is TRUE, but the rest of the procedure remains unchanged.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+    }
+
+		# project onto the selected models and compute the prediction accuracy for the left-out point
+	  as.search <- relax
```
It's not, it's just that I found `relax` confusing and in other places we use `as.search` instead. I removed the renaming and kept `relax`.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+	k_fold <- .get_kfold_poc(refmodel, K, verbose, seed)
+
+	# check that k_fold has the correct form
+	# .validate_kfold(refmodel, k_fold, refmodel$nobs)
```
I guess it's not, but the k-fold implementation is something that I haven't worked a lot on, so it might be better to focus on it in a subsequent PR.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  }
+  spath_cv <- lapply(seq_along(list_cv), function(fold_index) {
+    fold <- list_cv[[fold_index]]
+    family_kl <- fold$refmodel$family
```
It was defined as `fam` in the current version, I used `family` but didn't change it in all places. Now I've renamed it everywhere to `family`. I think this renaming should be applied as well to `family_kl`, because the `kl` part is quite specific and it's possible that in future versions we play with other divergences or distances?

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+  # of type refmodel computed by init_refmodel) and index list 'test_points' that denotes which
+  # of the data points were left out for the corresponding fold.
+
+  ## TODO: pass cvfits for refmodel so that each fold has a different fit with
```
I would defer kfold to a next PR if you are fine with it?

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+      refmodel$predfun(fit, newdata=newdata)
+    proj_predfun <- function(fit, newdata=default_data)
+      refmodel$proj_predfun(fit, newdata=newdata)
+    refmod <- get_refmodel_poc(refmodel$cvfits[[i]],
```
Yes, renamed it to `refmodel` for consistency and clarity.

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,370 @@
+}
+
+
+.get_kfold_poc <- function(refmodel, K, verbose, seed) {
```
So, originally, the folds can come from two places:

1. either because the user has fitted K number of fits and passed that already, or,
2. because the user has passed a cvfits function to recover the fit of each fold from the original fit.

Here we should just retrieve the proper fit and compute the statistics so there's not much overhead involved. Maybe cvfits is more complicated if it requires fitting each time. But, as I said in a previous comment, I think this is deep enough for a next PR and don't mingle too much on this one?

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+library(stringr)
```
I'll think about removing the dependency, I think I only use `str_match`.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  response <- attr(tt, "response")
+  global_intercept <- attr(tt, "intercept") == 1
+
+  hier <- str_match(terms, "\\|")
```
Agreed. Many of the functions in this file came to life to handle formulas in the most general way that I was able to. There are so many different ways of specifying a formula that I'm sure this can break in trizillion ways, so this is something to keep in mind to make our management more robust in the future. As users come up with weird and complex formulas it will be easier to improve.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+library(stringr)
+
+extract_terms_response <- function(formula) {
```
fixed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  response <- attr(tt, "response")
+  global_intercept <- attr(tt, "intercept") == 1
+
+  hier <- str_match(terms, "\\|")
```
it seems it does! fixed and removed `stringr` dependency

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+    response <- NA
+
+  list(individual_terms=individual_terms,
+       int_terms=int_terms,
```
fixed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+       global_intercept=global_intercept)
+}
+
+extract_response <- function(response) {
```
If `NA` is passed, then the function returns `NA` as well.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+       global_intercept=global_intercept)
+}
+
+extract_response <- function(response) {
```
These functions can be a little bit tricky so I will explain it here as a reference for when we talk in person. 

When you call `terms(form)` the response is an unevaluated object, it's not a string. At some points in projpred the response is actually a list of responses, where we are basically fitting the same model to different draws from the posterior for the parameters. In this case we want to split the formula to individual formulas so that lmer can fit them. Therefore, we leverage the splitting of these terms to several functions. This is the one that takes an unevaluated response object and returns a vector of strings, or a single string if it represents just a single response.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  return(response_name_ch)
+}
+
+validate_response_formula <- function(formula) {
```
As per my previous response, we don't really have multi response model as such, we just fit the model to several draws from the posterior.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+}
+
+flatten_individual_terms <- function(terms) {
+  if (length(terms) < 1)
```
Yes, NULL should also work, or rather, empty list.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+flatten_group_terms <- function(terms) {
+  if (length(terms) < 1)
+    return(terms)
+  split_terms <- strsplit(terms, "[ ]*\\|[ ]*")
```
nice thanks

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+    return(individual_terms)
+
+  if (length(group_terms) > 0 && return_group_terms)
+    warning(paste0("The model involves group terms.",
```
I commented it out for the moment.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  ## strong heredity by default
+  variables <- unlist(strsplit(term, ":"))
+  individual_joint <- paste(variables, collapse = " + ")
+  joint_intercept <- paste(c(individual_joint, term), collapse = " + ")
```
It seems a typo, renamed to `joint_term`.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+               gsub("^\\(", "",
+                    flatten_group_terms(term)))
+  if ("\\-" %in% term)
+    stop(paste0("Use of `-` is not supported, omit variables or use the ",
```
fixed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  tt <- extract_terms_response(formula)
+  response_name <- tt$response
+
+  X <- data[, colnames(data) != response_name]
```
We cannot, but we don't really care. The only thing we care about at this point is removing the response. lme4 will take care of fitting only the predictors that appear in the formula.

AlejandroCatalina replied:
refactored and simplified into two functions

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  tt <- extract_terms_response(formula)
+  response_name <- tt$response
+
+  X <- data[, colnames(data) != response_name]
```
This would break. Nonetheless, that case is not possible because this function is only used internally and the user has to give a reference model with a response (we cannot project without predictions about something), so we always have that response when we call `subset_formula`. But, it's a good point to make extra sure.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+
+  if (!is.null(ncol(y)) && ncol(y) > 1) {
+    ## remove parens from response
+    response_name <- gsub("\\(", "\\.",
```
We want to parse a vector of responses. The problem is that the response object as taken out of `terms(...)` is an unevaluated object that can look as `cbind(y.1, y.2)`. So to parse it we need to remove "cbind" and the brackets.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  return(length(group_terms) > 0)
+}
+
+subset_formula <- function(formula, terms, data, y=NULL, return_list=FALSE) {
```
Okay, maybe `subset_submodel`.

AlejandroCatalina replied:
after discussing we'll break this up into 2 functions

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  return(length(group_terms) > 0)
+}
+
+subset_formula <- function(formula, terms, data, y=NULL, return_list=FALSE) {
```
Okay, it can be more verbose. I'll rename the argument to `split_formulas`. The intention of the argument is to return a list of formulas where each of them has a single response, instead of a formula with multiple responses.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  response <- attr(tt, "response")
+  variables.list <- as.list(attr(tt, "variables")[-1])
+
+  if (response)
```
So, this follows from the same issue as the responses. When we run `as.list(attr(tt, "variables"))` the first element of the list is "list", so we need to remove that. Then, the first element of the remaining list is the response, if there is response.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  term.labels <- attr(tt, "term.labels")
+  intercept <- attr(tt, "intercept")
+
+  mm <- model.matrix(formula, data)
```
No, that's a good point. Right now, it is intended to break if group terms are passed. This function is a utility to split a matrix into individual predictors in the formula. For instance, `y ~ x`, where `x` is a matrix of 20 predictors, would be split as `y ~ x1 + ... + x20`. But it indeed does not work with group terms.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  list(formula=formula, data=new_data)
+}
+
+split_linear_term <- function(term, term_formula, data) {
```
It is tied to `break_up_matrix_term`. These are ad-hoc functions to break matrix terms in a formula when they represent a matrix of linear predictors.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  tt <- extract_terms_response(submodel)
+  ind_int_terms <- length(tt$individual_terms) + length(tt$int_terms)
+  group_terms <- sum(unlist(lapply(tt$group_terms, count_terms_in_group_term)))
+  if (is.null(group_terms))
```
Right, I don't think `sum` can return `NULL`. Removed this if because it's redundant in that case.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  term <- gsub("[\\(\\)]", "", flatten_group_terms(term))
+
+  chunks <- strsplit(term, " \\| ")[[1]]
+  variables <- strsplit(chunks[1], " \\+ ")[[1]]
```
Refactored to use `extract_terms_response`.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+}
+
+count_terms_in_submodel <- function(submodel) {
+  if (class(submodel) != "formula")
```
fixed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  split_term
+}
+
+count_terms_in_submodel <- function(submodel) {
```
We only support formulas as input, but internally there are stages where we have to handle strings for some things.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  split_term
+}
+
+count_terms_in_submodel <- function(submodel) {
```
renamed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+
+sort_submodels_by_size <- function(submodels) {
+  size <- lapply(submodels, count_terms_in_submodel)
+  df <- as.data.frame(cbind(submodels=submodels, size=unlist(size)))
```
It's a list of formulas for different submodels.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+
+  ord_list <- groups
+  ord_list_nona <- lapply(ord_list, function(l) l[!is.na(l)])
+  ord_list_nona[!is.na(ord_list_nona)]
```
Not really, in the first line we remove the NAs inside the elements of the list that are not NA already (those return NA again), the second line removes the NA elements from the overall list.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  ord_list_nona[!is.na(ord_list_nona)]
+}
+
+count_variables_chosen <- function(refmodel, list_of_terms) {
```
Yes. The problem that I encountered is that counting variables is not just the length of the list_of_terms, we need to take into account the intercepts and whatnot, so we have to come back to the actual model. I'll try to see if there's a simpler way.

AlejandroCatalina replied:
after discussing we'll try to simplify it

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  count_terms_in_submodel(flatten_formula(form))
+}
+
+is.redundant <- function(refmodel, current, new) {
```
renamed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  count_terms_in_submodel(flatten_formula(form))
+}
+
+is.redundant <- function(refmodel, current, new) {
```
renamed twice. We are checking if the submodel after adding `new` predictor is redundant with the current submodel. This can happen as when we sequentially increase the models some of the predictors may be included in a previous step.

AlejandroCatalina commented on R/kl_helpers.R:
```diff
@@ -6,18 +6,51 @@
 # Missing: Quasi-families not implemented. If dis_gamma is the correct shape
 # parameter for projected Gamma regression, everything should be OK for gamma.
 
+## TODO: rename this, kl_helper does not help anybody
```
`extend_family` maybe?

AlejandroCatalina commented on R/kl_helpers.R:
```diff
@@ -6,18 +6,51 @@
+## TODO: rename this, kl_helper does not help anybody
+#' @export
 kl_helpers <- function(fam) {
+  if (.has.fam.extras(fam))
```
changed

AlejandroCatalina commented on R/kl_helpers.R:
```diff
@@ -6,18 +6,51 @@
+
+kl_helpers_binomial <- function(fam) {
+  kl_dev <- function(pref, data, psub) {
+    if(NCOL(pref$mu)>1) {
```
done here and in other places

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+#' of ones.
+#' @param transform Should the linear predictor be transformed using the
+#' inverse-link function? Default is \code{FALSE}. For \code{proj_linpred} only.
+#' @param integrated If \code{TRUE}, the output is averaged over the
```
It averages over the samples from the projected posterior.

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+#' inverse-link function? Default is \code{FALSE}. For \code{proj_linpred} only.
+#' @param integrated If \code{TRUE}, the output is averaged over the
+#' parameters. Default is \code{FALSE}. For \code{proj_linpred} only.
+#' @param nv Number of variables in the submodel (the variable combination is
```
okay, I guess it applies to `nv`, `nc` and `ns` particularly?

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+#' @param draws Number of draws to return from the predictive distribution of
+#' the projection. The default is 1000.
+#' For \code{proj_predict} only.
+#' @param seed_samp An optional seed to use for drawing from the projection.
```
renamed

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+                 ' varsel, cv_varsel or project'))
+  }
+
+  ## FIXME: proj$sub_fit always exists, but does $coefficients?
```
Not right now, but if we include projections onto gaussian processes, for instance, we will have to worry about this, because not every projection model has coefficients. Right now, as we are doing linear projections, we always have coefficients. It also largely depends on the implementation of the model, retrieving the attribute by name seems very unsafe to me, as lme4 may decide to change the name of their attributes.

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+               'number of variable indices (vind).'))
+
+  # set random seed but ensure the old RNG state is restored on exit
+  rng_state_old <- rngtools::RNGseed()
```
Honestly, I don't really know, I carried this from the original code. I'll look it up.

AlejandroCatalina replied:
@jpiironen do you have any insight on this?

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+    if (integrated) {
+      # average over the parameters
+      pred <- as.vector( proj$weights %*% pred )
+    } else if (!is.null(dim(pred)) && dim(pred)[1]==1) {
```
I guess what we are saying is that if the user does not want the integrated behaviour but only has one prediction, work on it as a vector. Otherwise work with it as a matrix. Maybe the matrix behaviour holds in this case as well, I'll look it up.

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+  # function to perform to each projected submodel
+  fun <- function(proj, mu, offset, weights) {
+    pred <- t(mu)
+    if (!transform) pred <- proj$family_kl$linkfun(pred)
```
Yes I agree. I think we could leverage this entirely to the extended family, where we have all the information on how to output (transformed) predictions.

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+      }
+      list(pred = pred, lpd = lpd)
+    } else {
+      pred
```
fixed

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+  }
+
+  # proj_helper lapplies fun to each projection in object
+  proj_helper_poc(object = object, xnew = xnew, offsetnew = offsetnew,
```
fixed

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+
+  # function to perform to each projected submodel
+  fun <- function(proj, mu, offset, weights) {
+    if(is.null(draws)) draws <- 1000
```
fixed as default for the argument on the function prototype

AlejandroCatalina commented on R/mle_helpers.R:
```diff
@@ -0,0 +1,98 @@
+library(lme4)
```
fixed

AlejandroCatalina commented on R/mle_helpers.R:
```diff
@@ -0,0 +1,98 @@
+library(penalized)
+library(optimx)
+
+fetch_data <- function(data, data_points=NULL, newdata=NULL) {
```
renamed to `obs`. `newdata` refers to prediction data or held-out data, not to the data points to index.

AlejandroCatalina commented on R/mle_helpers.R:
```diff
@@ -0,0 +1,98 @@
+    return(newdata[data_points,, drop = FALSE])
+}
+
+linear_mle <- function(form, dat, regul=NULL)
```
argument `regul` is there to have the same interface in all of these functions so that the caller does not need to care about it. I'll probably delete the penalised method as we don't really use it.

AlejandroCatalina commented on R/mle_helpers.R:
```diff
@@ -0,0 +1,98 @@
+                           standardize=TRUE) # standard l2 regularisation
+    }
+  }
+  if (class(form) == "formula")
```
fixed

AlejandroCatalina commented on R/mle_helpers.R:
```diff
@@ -0,0 +1,98 @@
+    all <- extract_terms_response(f)
+    response <- dat[, as.character(all$response)]
+    ## no group terms allowed for penalized MLE
+    covariates <- c(all$individual_terms, all$int_terms)
```
No need, this is just for the penalisation part. The group terms cannot be penalised in this framework, they just go as MLE predictors, but I don't really like the library, I think we should just remove it.

AlejandroCatalina commented on R/mle_helpers.R:
```diff
@@ -0,0 +1,98 @@
+  else if (class(form) == "list")
+    lapply(form, function(f) fit_penalized(f))
+  else
+    stop("The provided formula is neither a formula nor a list")
```
fixed

AlejandroCatalina commented on R/mle_helpers.R:
```diff
@@ -0,0 +1,98 @@
+    return(newdata[data_points,, drop = FALSE])
+}
+
+linear_mle <- function(form, dat, regul=NULL)
```
fixed

AlejandroCatalina commented on R/mle_helpers.R:
```diff
@@ -0,0 +1,98 @@
+    stop("The provided formula is neither a formula nor a list")
+}
+
+multilevel_mle <- function(form, dat, regul=NULL) {
```
fixed

AlejandroCatalina commented on R/project_formula.R:
```diff
@@ -0,0 +1,123 @@
+#' proj4 <- project(vs, nv = 4)
+#'
+#' # project onto an arbitrary variable combination (variable indices 3,4 and 8)
+#' proj <- project(fit, vind=c(3,4,8))
```
I think we could support names, although they're also prone to mistakes by the user.

AlejandroCatalina commented on R/project_formula.R:
```diff
@@ -0,0 +1,123 @@
+
+  if (is.null(relax))
+  	# use non-relaxed solution for datafits by default
+    relax <- ifelse('datafit' %in% class(get_refmodel(object)), FALSE, TRUE)
```
fixed

AlejandroCatalina commented on R/project_formula.R:
```diff
@@ -0,0 +1,123 @@
+  }
+
+  if (is.null(ns) && is.null(nc))
+    ns <- min(400, NCOL(refmodel$mu)) # by default project at most 400 draws
```
fixed

AlejandroCatalina commented on R/project_formula.R:
```diff
@@ -0,0 +1,123 @@
+
+  if (is.null(nv)) {
+    if (!is.null(object$ssize) && !is.na(object$ssize))
+      nv <- object$ssize # by default, project onto the suggested model size
```
renamed to `suggested_size`.

AlejandroCatalina commented on R/project_formula.R:
```diff
@@ -0,0 +1,123 @@
+
+	# project onto the submodels
+	if (relax) {
+	  subm <- .get_submodels_poc(list(vind=vind), nv, family_kl, p_ref,
```
We don't really handle the intercept independently, it's just an indicator of whether to fit an intercept in the projections or not.

AlejandroCatalina commented on R/project_formula.R:
```diff
@@ -0,0 +1,123 @@
+#' Default is the model size suggested by the variable selection (see function \code{suggest_size}).
+#'  Ignored if \code{vind} is specified.
+#' @param vind Variable indices onto which the projection is done. If specified, \code{nv} is ignored.
+#' @param relax If TRUE, then the projected coefficients after L1-selection are computed
```
I agree that the documentation it's a bit confusing, I'll try to reword it. Fixed the NULL thing.

AlejandroCatalina commented on R/project_formula.R:
```diff
@@ -0,0 +1,123 @@
+	}
+
+	# add family_kl
+	proj <- lapply(subm, function(model) {
```
I see why it would be confusing. I'll try to simplify it.

AlejandroCatalina commented on R/projfun_formula.R:
```diff
@@ -0,0 +1,122 @@
+  kl <- family_kl$kl(ref, list(weights=wobs), list(mu=musub, dis=dis_sub))
+  submodel <- list(kl = kl, dis = dis_sub, weights = wsample)
+
+  ## split b to alpha and beta, add it to submodel and return the result
```
Yes.

AlejandroCatalina commented on R/projfun_formula.R:
```diff
@@ -0,0 +1,122 @@
+        sub_refit <- searchpath$sub_fits[[nv]]
+      }
+
+      if (family_kl$family == "gaussian")
```
Because we work with the pseudo-Gaussian transformation of the log likelihoods, the required information to compute the dispersion and the KL divergence differ by family, and for the Gaussian specifically.

AlejandroCatalina commented on R/projfun_formula.R:
```diff
@@ -0,0 +1,122 @@
+    project_submodel_poc(vind, p_ref, refmodel, family_kl, intercept, regul=regul))
+}
+
+.get_submodels_poc <- function(searchpath, nv, family_kl, p_ref,
```
It's a list containing the order in which the variables were chosen.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -61,36 +62,38 @@ get_refmodel.cvsel <- function(object, ...) {
+	if ('lmerMod' %in% class(object) || 'gamm4' %in% class(object))
+    x <- as.matrix(object$x)
+  else {
+    x <- rstanarm::get_x(object)
```
Agreed, this is carried from the original codebase and it's not used anymore except for backwards compatibility.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -61,36 +62,38 @@ get_refmodel.cvsel <- function(object, ...) {
 #' @rdname get-refmodel
 #' @export
 get_refmodel.stanreg <- function(object, ...) {
```
Agreed.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -207,46 +210,46 @@ get_refmodel.brmsfit <- function(object, ...) {
   }
-  
+
   init_refmodel(
```
Agreed. Right now we cannot unless we directly pull out compatibility with the current version, which could be a bit drastic. We should maybe wait a bit?

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -267,85 +270,92 @@ get_refmodel.brmsfit <- function(object, ...) {
 	family <- kl_helpers(family)
-	
+
+  mu_fun <- function(x, alpha, beta, offset) {
```
This is the old version and not used anymore.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -440,37 +450,210 @@ init_refmodel <- function(z, y, family, x=NULL, predfun=NULL, dis=NULL, offset=N
 
 #' @export
 predict.refmodel <- function(object, znew, ynew = NULL, offsetnew = NULL,
-                             weightsnew = NULL, type = c('response', 'link'), ...) {
```
So, predict can be either the linear predictor or the posterior predict, so it's not entirely clear if we should rename it to posterior_predict.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -440,37 +450,210 @@ init_refmodel <- function(z, y, family, x=NULL, predfun=NULL, dis=NULL, offset=N
+get_refmodel_poc.default <- function(fit, data, y, formula, predfun, proj_predfun,
+                                     mle, fetch_data, family=NULL, wobs=NULL,
+                                     folds=NULL, cvfits=NULL, penalized=FALSE) {
+  fetch_data_wrapper <- function(data_points=folds, newdata=NULL)
```
Yes. We can work on making it more explicit.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -440,37 +450,210 @@ init_refmodel <- function(z, y, family, x=NULL, predfun=NULL, dis=NULL, offset=N
+}
+
+#' export
+get_refmodel_poc.brmsfit <- function(fit, data=NULL, y=NULL, formula=NULL,
```
Agreed.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -440,37 +450,210 @@ init_refmodel <- function(z, y, family, x=NULL, predfun=NULL, dis=NULL, offset=N
+  terms <- extract_terms_response(formula)
+  if (is.null(predfun))
+    predfun <- function(fit, newdata=NULL)
+      t(posterior_linpred(fit, transform=TRUE, newdata=newdata))
```
For some reason we work with NxS (or the other way around?), where N is number of datapoints and S is number of draws.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -440,37 +450,210 @@ init_refmodel <- function(z, y, family, x=NULL, predfun=NULL, dis=NULL, offset=N
+    predfun <- function(fit, newdata=NULL)
+      t(posterior_linpred(fit, transform=TRUE, newdata=newdata))
+
+  if (is.null(mle) && is.null(proj_predfun))
```
I'm not sure. I think the safest option is that you either pass all of them or none of them. For most use cases the default, provided functions should be enough, tho, so the user does not need to pass anything.

AlejandroCatalina commented on R/refmodel.R:
```diff
@@ -440,37 +450,210 @@ init_refmodel <- function(z, y, family, x=NULL, predfun=NULL, dis=NULL, offset=N
+        proj_predfun <- penalized_linear_proj_predfun
+
+  fetch_data_wrapper <- function(data_points=folds, newdata=NULL)
+    as.data.frame(fetch_data(data, data_points, newdata))
```
It's more of a way to ensure that it is a data frame all the way, because some methods behave differently with a data frame and with a matrix.

AlejandroCatalina commented on R/search_formula.R:
```diff
@@ -0,0 +1,84 @@
+
+search_L1_poc <- function(p_ref, refmodel, family, intercept, nv_max, penalty, opt) {
+  terms_ <- break_formula(refmodel$formula)
+  x <- model.matrix(refmodel$formula, refmodel$fetch_data())
```
It does not, exactly. Look at Juho's comment earlier in this thread, that's something we have to work on.

AlejandroCatalina commented on R/summary_funs_formula.R:
```diff
@@ -0,0 +1,33 @@
+.get_sub_summaries_poc <- function(submodels, test_points, refmodel, family_kl, groups=NULL) {
```
Right now there would a conflict of names in the package, so I kept them to avoid removing the other files directly, but I agree that we should decide something better, either drop the current code (so we don't have duplicated code) or to choose a more elegant suffix.

AlejandroCatalina commented on R/varsel_formula.R:
```diff
@@ -0,0 +1,162 @@
+  groups <- args$groups
+  group_features <- formula_contains_group_terms(refmodel$formula)
+
+  if (tolower(method) == 'l1' && group_features) {
```
fixed

AlejandroCatalina commented on R/varsel_formula.R:
```diff
@@ -0,0 +1,162 @@
+                       nspred = NULL, ncpred = NULL, relax=FALSE, nv_max = NULL,
+                       intercept = NULL, penalty=NULL, verbose = F,
+                       lambda_min_ratio=1e-5, nlambda=150, thresh=1e-6, regul=1e-4,
+                       groups=NULL, ...) {
```
It's a way for the user to pass a specific list of terms to consider for projection instead of the default way of breaking the formula.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+                  unlist(lapply(int_terms, split_interaction_term)))
+
+  ## exclude the intercept if there is no intercept in the reference model
+  if (!global_intercept) {
```
after discussing this should be fine but we have to test this

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  joint_intercept
+}
+
+split_group_term <- function(term) {
```
after discussing updated

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  for (size in unique(ordered$size))
+    groups[[size]] <- as.character(ordered$submodels[ordered$size == size])
+
+  ord_list <- groups
```
undo rename

AlejandroCatalina commented on R/methods_formula.R:
```diff
@@ -0,0 +1,204 @@
+#' corresponds to the number trials per observation. For \code{proj_linpred},
+#' this argument matters only if \code{ynew} is specified. By default a vector
+#' of ones.
+#' @param transform Should the linear predictor be transformed using the
```
this has to be discussed at a more general level

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,422 @@
+  term <- gsub("\\)$", "",
+               gsub("^\\(", "",
+                    flatten_group_terms(term)))
+  if ("\\-" %in% term)
```
fixed

AlejandroCatalina commented on R/kl_helpers.R:
```diff
@@ -6,18 +6,51 @@
+         'Student_t' = kl_helpers_student_t(fam))
+}
+
+kl_helpers_binomial <- function(fam) {
```
deferred to a later stage
we will also leverage some more functionality regarding mu computation to the families

AlejandroCatalina commented:
Okay, just pushed some commits addressing all the comments and discussions we've had. 

There are some big things like removing the current codebase and copying some unchanged code to the new files. This is because in anyway we are going to mark the current version with a safe milestone so people will have access to it as a stable release anyway. This should also avoid some burden on this PR and in general because having two codebases is quite prone to errors. 

I also added a lot of documentation to formula.R which is the big new player, so that should improve readability of what's going on there.

Something on the TODO list is to reframe the tests because right now Travis fails basically because I've removed the code being tested, so we have to fix the tests.

In short, I think this is ready for the second round of polishing!

paul-buerkner commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+sort_submodels_by_size <- function(submodels) {
+  size <- lapply(submodels, count_terms_in_subformula)
+  df <- as.data.frame(cbind(submodels=submodels, size=unlist(size)))
+  df$size <- as.numeric(df$size)
```
Can we avoid creating a matrix and then a data.frame and create a data.frame directly? Then we can also get rid of the `as.numeric`.

AlejandroCatalina replied:
fixed

paul-buerkner commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+count_variables_chosen <- function(formula, list_of_terms) {
+  if (length(list_of_terms) == 0)
+    return(0)
+  formula <- subset_formula(formula, unique(unlist(list_of_terms)))
```
since subset_formula replaces all terms currently, do we need the reference formula here at all?

AlejandroCatalina replied:
true, seems that we can do without it!

paul-buerkner commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+is_next_submodel_redundant <- function(formula, current, new) {
+  old_submodel <- current
+  new_submodel <- c(current, new)
+  if (count_variables_chosen(formula, new_submodel) >
```
I know we discussed this, but how can we be sure some model is redundant just because it has less terms? After all, it could contain *different* terms?

paul-buerkner commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+#' @return a vector of incremental non redundant submodels for all the possible terms
+#' included.
+reduce_models <- function(formula, chosen) {
+  Reduce(function(chosen, x)
```
can we split this up into a few more readable calls instead of one big one?

AlejandroCatalina commented on R/cv_varsel_formula.R:
```diff
@@ -0,0 +1,488 @@
+#' Cross-validate the variable selection (varsel)
```
fixed

AlejandroCatalina commented on R/extend_family.R:
```diff
@@ -0,0 +1,202 @@
+# parameter for projected Gamma regression, everything should be OK for gamma.
+
+#' @export
+extend_family <- function(fam) {
```
added documentation

AlejandroCatalina commented on R/extend_family.R:
```diff
@@ -0,0 +1,202 @@
+# parameter for projected Gamma regression, everything should be OK for gamma.
+
+#' @export
+extend_family <- function(fam) {
```
renamed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+  nlist(individual_terms,
+        interaction_terms,
+        group_terms,
+        response=extract_response(response),
```
done!

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+#' @param response The response as retrieved from the formula object.
+#' @return the response as a character vector.
+extract_response <- function(response) {
+  response_name_ch <- as.character(response[[1]])
```
checked the length, I may come up with a better way, but so far at least we accept a single element list.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+  split_terms <- strsplit(terms, "[ ]*\\|([^\\|]*\\||)[ ]*")
+  group_names <- unique(unlist(lapply(split_terms, function(t) t[2])))
+
+  groups <- setNames(lapply(group_names, function(g)
```
renamed to group_terms, yes

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+    group_names)
+  groups <- lapply(groups, function(g) unlist(g[!is.na(g)]))
+
+  groups <- lapply(1:length(groups), function(i) {
```
nice!

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+  split_terms <- strsplit(terms, "[ ]*\\|([^\\|]*\\||)[ ]*")
+  group_names <- unique(unlist(lapply(split_terms, function(t) t[2])))
+
+  groups <- setNames(lapply(group_names, function(g)
```
yes it does (more test for these kind of weirdly specified models are needed anyway)

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+      t.labels <- c("1")
+
+    if (!attr(t, "intercept"))
+      paste0("(0 + ", paste(t.labels, collapse=" + "), " | ", g_name, ")")
```
fixed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+}
+
+#' Simplify and flatten group terms by breaking them down in as many terms
+#' as varying effects. It also explicitly adds or removes the varying intercept.
```
that was confusing indeed, this function only flattens group terms, meaning that it unifies all terms under the same group, so this was totally incorrect documentation for this function. In the split_group_terms function, though, we do split your example in the given terms plus adding the overall effects.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+  return(unlist(groups))
+}
+
+#' Simplify and flatten group terms by breaking them down in as many terms
```
yes!

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+#' @param formula A formula for a valid model.
+#' @param return_group_terms If TRUE, return group terms as well. Default TRUE.
+#' @return a vector of all the minimal valid terms that make up for submodels.
+break_formula <- function(formula, return_group_terms=TRUE) {
```
renamed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+
+  ## if there are no group level terms then `terms` is the most basic
+  ## decomposition already
+  if (length(group_terms) == 0 && length(interaction_terms) == 0)
```
I think it was not necessary after all, if the others are empty the if and lapply that follows should leave them empty.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+
+  chunks <- strsplit(term, "[ ]*\\|([^\\|]*\\||)[ ]*")[[1]]
+  lhs <- as.formula(paste0("~", chunks[1]))
+  tt <- extact_terms_response(lhs)
```
fixed

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+                            function(v)
+                              paste0(split_interaction_term(v), " + ", "(1 | ", group, ")")))
+  } else {
+    group_terms <- lapply(lin_v, function(v) paste0(v, " + ", "(0 + ", v, " | ", group, ")"))
```
`split_interaction_term` also returns the interaction. So, "x1:x2" is split as "x1 + x2 + x1:x2", and that is added to the group term. My question is, should we add the overall effects as varying effects as well? Right now we are only adding the interaction at the group level.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+  response_name <- tt$response
+
+  if (!is.null(ncol(y)) && ncol(y) > 1) {
+    response_cols <- unlist(lapply(1:ncol(y), function(n)
```
neat!

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+    }
+    data <- data.frame(y, data[, colnames(data) != response_name])
+    colnames(data)[1:ncol(y)] <- response_cols
+  }
```
done!

AlejandroCatalina commented on R/formula.R:
```diff
@@ -0,0 +1,407 @@
+#' @return the number of terms in the subformula.
+count_terms_in_subformula <- function(subformula) {
+  if (!inherits(subformula, "formula"))
+    subformula <- as.formula(paste0("~ ", subformula))
```
Yeah, `subformula` is expected to be either a formula already or the right hand side as a string.

AlejandroCatalina commented on R/formula.R:
```diff
@@ -91,7 +94,7 @@ flatten_formula <- function(formula) {
 #' @return a vector of unique linear individual terms.
 flatten_individual_terms <- function(terms) {
-  if (is.null(terms))
+  if (length(terms) == 0)
```
Interestingly, `NULL` does not work here because for some reason `terms` might be `character(0)`. I'll investigate this.

AlejandroCatalina commented:
After the last discussion with @paul-buerkner , this should be good to go into develop. There are still couple issues to deal with in next PRs before merging into master:

1. deal with `d_test`,

jpiironen commented:
I don't have other comments, but please try making sure all the old tests pass (at least those functionalities that are supported in the new version). In particular, I would like to make sure the `test_datafit` and `test_glm_elnet` work fine, to make sure the L1-search works fine after the changes (cf. my comment earlier).

AlejandroCatalina commented:
Yes, I will wrap a method around alpha and beta to respect the current L1 method. I want to understand nonetheless why glmnet would be different. Test_glm_elnet should work exactly the same as we didn’t touch any of those functions. Test_datafit May not work yet but it’s on the todo list before merging into master for sure. Thanks!

Best,
Alejandro
On 26 Feb 2020, 11:21 PM +0200, Juho Piironen <notifications@github.com>, wrote:
> I don't have other comments, but please try making sure all the old tests pass (at least those functionalities that are supported in the new version). In particular, I would like to make sure the test_datafit and test_glm_elnet work fine, to make sure the L1-search works fine after the changes (cf. my comment earlier).
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

paul-buerkner commented:
Ok, this PR is ready to be merged into 'develop'. Before merging into master, there is a lot we need to check, fix, and clean up but this will be handled in a new PR. Thank you @AlejandroCatalina for all the hard work on this!
